lib/sequent/core/aggregate_root.rb

```diff
@@ -98,12 +98,37 @@ def to_s
         "#{self.class.name}: #{@id}"
       end
 
+      # Some aggregates represent a unique external entity (e.g. a
+      # user's email address or login name) and this uniqueness needs
+      # to be enforced. For each unique key the returned object should
+      # have an entry where the key of the entry describes the scope
+      # of the constraint (e.g. `user_email` or `login_name`) and the
+      # value represents the unique value. Values can be any JSON
+      # value (string, object, array, etc). Note that uniqueness is
+      # enforced across all aggregate types if the same scope is used.
+      #
+      # An `AggregateKeyNotUniqueError` is raised if a unique
+      # constrained is violated when committing the events to the
+      # database.
+      #
+      # Sample return value:
+      #
+      # ```
+      # {
+      #   user_email: { email: 'bob@example.com' }
```

Comment:
I prefer something more descriptive like:

```
class User < AggregateRoot
  unique_constraint :email, scope: :user_email
  unique_constraint :first_name, :last_name, scope: :user_name
end
```

This will add two unique constraints.

The `email`, `first_name` and `last_name` are then methods/attr_readers on the aggregate that will called when committing the events to generate the internal structure.

Reply:
The `key` can also be an array of values instead of a hash.

Reply:
Should order matter? In other words are:

`unique_constraint :first_name, :last_name, scope: :user_name` and `unique_constraint :last_name, :first_name, scope: :user_name` the same?

Reply:
They are the same (we use a JSONB column to store the key and PostgreSQL normalizes objects by field name, etc). Arrays would be different though...